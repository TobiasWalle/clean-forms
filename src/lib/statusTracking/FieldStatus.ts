export interface FieldStatusArguments {
  dirty: boolean;
  touched: boolean;
}

export class FieldStatus {
  public readonly dirty: boolean;
  public readonly pristine: boolean;
  public readonly touched: boolean;
  public readonly untouched: boolean;

  constructor(args: FieldStatusArguments) {
    const { dirty, touched } = args;

    this.dirty = dirty;
    this.pristine = !dirty;
    this.touched = touched;
    this.untouched = !touched;
  }
}

export const DEFAULT_FIELD_STATUS: FieldStatus = new FieldStatus({ dirty: false, touched: false });

export function cloneFieldStatus<Children = any>(statusToClone: FieldStatus, overrideArgs: Partial<FieldStatusArguments>): FieldStatus {
  return new FieldStatus({
    ...(statusToClone as any),
    ...overrideArgs
  });
}
