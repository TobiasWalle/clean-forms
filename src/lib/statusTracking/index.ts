export * from './FieldStatus';
export * from './FieldStatusMapping';
