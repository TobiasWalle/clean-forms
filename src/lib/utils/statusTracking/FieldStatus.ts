export interface FieldStatus {
  valid: boolean;
  inValid: boolean;
  pristine: boolean;
  dirty: boolean;
  untouched: boolean;
  touched: boolean;
  error?: string;
}

export const DEFAULT_FIELD_STATUS: FieldStatus = {
  valid: true,
  inValid: false,
  pristine: true,
  dirty: false,
  untouched: true,
  touched: false
};
