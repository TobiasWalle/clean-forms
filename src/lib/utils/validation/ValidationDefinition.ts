export type False = '0';
export type True = '1';
export type If<C extends True | False, Then, Else> = { '0': Else, '1': Then }[C];

export type Diff<T extends string, U extends string> = (
  { [P in T]: P } & { [P in U]: never } & { [x: string]: never }
)[T];

export type X<T> = Diff<keyof T, keyof Object>;

export type Is<T, U> = (Record<X<T & U>, False> & Record<any, True>)[Diff<X<T>, X<U>>];

type DeepPartial<T> = {
    [P in keyof T]?: If<Is<Function & T[P], Function>, T[P], DeepPartial<T[P]>>
}

export interface ValidationFunctionArgs<Value = any, Model = any> {
  value: Value;
  formValue: Model;
}

export type ValidationFunction<Value = any, Model = any> =
  (args: ValidationFunctionArgs<Value, Model>) => string | null;

export class ArrayValidation<Model = any, Item = any, ArrayType = Item[]> {
  constructor(
    public readonly itemValidation: ValidationFunction<Item, Model> | ValidationDefinition<Item> | ArrayValidation<Model> | null,
    public readonly arrayValidation?: ValidationFunction<ArrayType, Model>
  ) {
  }
}

export type ValidationResolver<Value = any, Model = any> = ArrayValidation<Model, any, Value> | ValidationFunction<Value, Model>;

export interface ValidationDefinition<Model> {
  [key: string]: ValidationResolver<any, any> | ValidationDefinition<any> ;
}
