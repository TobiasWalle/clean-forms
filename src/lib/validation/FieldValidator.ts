import {
  asPath,
  combinePaths,
  getPathAsString,
  Path,
  path as createPath,
} from '../models/Path';
import { isYupSchema } from '../utils/isYupSchema';
import { selectDeep } from '../utils/selectDeep';
import { getValidationDefinitionPaths } from './getValidationDefinitionPaths';
import { selectDeepValidator } from './selectDeepValidator';
import {
  ValidationDefinition,
  ValidationError,
  ValidationErrors,
  ValidationFunction,
} from './ValidationDefinition';

interface ValidateFieldArguments<Model> {
  model: Model;
  validationDefinition: ValidationDefinition<Model>;
  path: Path<Model>;
}

interface ValidateFieldWithYupArguments<Model> {
  model: Model;
  validationDefinition: import('yup').Schema<Model>;
  path: Path<Model>;
}

export interface ValidateModelArguments<Model> {
  model: Model;
  validationDefinition: ValidationDefinition<Model>;
}

export type FieldError = string | undefined;

type FieldErrors<Model> = Array<[Path<Model>, FieldError]>;

export interface FieldErrorMapping {
  [path: string]: FieldError;
}

export function combineValidationDefinitions<Value>(
  ...validators: Array<ValidationDefinition<Value>>
): ValidationFunction<Value> {
  return value => {
    return validators.reduce(
      (allErrors, validationDefinition) => {
        const validationResult: ValidationErrors = Object.entries(
          validateModel({
            model: value,
            validationDefinition,
          })
        ).map(([path, error]) => [path, error || null]);
        return [...allErrors, ...validationResult];
      },
      [] as ValidationErrors
    );
  };
}

export function validateModel<Model = any>({
  model,
  validationDefinition,
}: ValidateModelArguments<Model>): FieldErrorMapping {
  let errors: Array<Array<readonly [string, FieldError]>> = [];
  if (isYupSchema(validationDefinition)) {
    errors = [
      validateFieldWithYup({
        model,
        validationDefinition,
        path: createPath(),
      }).map(([path, error]) => [getPathAsString(path), error] as const),
    ];
  } else {
    errors = getValidationDefinitionPaths(validationDefinition, model).map(
      path => {
        const fieldErrors = validateField({
          model,
          validationDefinition,
          path,
        });
        return fieldErrors.map(
          ([errorPath, error]) =>
            [getPathAsString(combinePaths(path, errorPath)), error] as const
        );
      }
    );
  }

  const result: FieldErrorMapping = {};
  errors.forEach(arr =>
    arr.forEach(([path, error]) => {
      if (error != null) {
        result[path] = error;
      }
    })
  );
  return result;
}

function validateField<Model>(
  args: ValidateFieldArguments<Model>
): FieldErrors<Model> {
  const { model, validationDefinition, path } = args;
  const value = selectDeep({ object: model, path, assert: false });
  const validation = selectDeepValidator({
    object: validationDefinition,
    path,
    assert: false,
  });
  if (!validation) {
    return [];
  } else if (isYupSchema(validation)) {
    return validateFieldWithYup({
      model,
      validationDefinition: validation,
      path,
    });
  } else {
    return runValidationFunctionInTryCatchAndCheckType(
      path,
      value,
      model,
      validation
    );
  }
}

function validateFieldWithYup<Model>({
  model,
  validationDefinition,
  path,
}: ValidateFieldWithYupArguments<Model>): FieldErrors<Model> {
  try {
    const value = selectDeep({ object: model, path, assert: false });
    validationDefinition.validateSync(value, {
      abortEarly: false,
      strict: true,
    });
  } catch (err) {
    const yupError = err as import('yup').ValidationError;
    return yupError.inner.map(innerError => [
      asPath(innerError.path || ''),
      innerError.message,
    ]);
  }
  return [];
}

function runValidationFunctionInTryCatchAndCheckType<Model>(
  path: Path<Model>,
  value: any,
  model: Model,
  validationFunction: any
): FieldErrors<Model> {
  if (typeof validationFunction === 'function') {
    return runValidationFunctionInTryCatch(
      path,
      value,
      model,
      validationFunction
    );
  } else {
    const pathAsString = getPathAsString(path);
    const errorMessage = `Invalid validation type "${typeof validationFunction}" for path "${pathAsString}"`;
    console.error(errorMessage);
    return [[createPath(), errorMessage]];
  }
}

function runValidationFunctionInTryCatch<Model>(
  path: Path<Model>,
  value: any,
  model: Model,
  validationFunction: ValidationFunction | undefined | null
): FieldErrors<Model> {
  try {
    return runValidationFunctionIfDefined(value, model, validationFunction);
  } catch (e) {
    value = JSON.stringify(value);
    const pathAsString = getPathAsString(path);
    const errorMessage = `Error while running validation function for value ${value} in path ${pathAsString}:`;
    console.error(errorMessage, e);
    return [[createPath(), errorMessage]];
  }
}

function runValidationFunctionIfDefined<Model>(
  value: any,
  model: Model,
  validationFunction: ValidationFunction | undefined | null
): FieldErrors<Model> {
  if (validationFunction == null) {
    return [];
  }
  const validationResult = validationFunction(value);
  if (validationResult instanceof Array) {
    return validationResult.map(([pathLike, error]): [
      Path<Model>,
      FieldError
    ] => [asPath(pathLike) as Path<Model>, validationErrorToFieldError(error)]);
  } else {
    return [[createPath(), validationErrorToFieldError(validationResult)]];
  }
}

function validationErrorToFieldError(
  validationError: ValidationError
): FieldError {
  return validationError || undefined;
}
