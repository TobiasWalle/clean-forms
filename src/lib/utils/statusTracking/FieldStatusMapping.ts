import { FieldStatus } from './FieldStatus';

export type FieldStatusMapping<Model> = {
  [key in keyof Model]?: FieldStatus | FieldStatusMapping<Model>;
};
