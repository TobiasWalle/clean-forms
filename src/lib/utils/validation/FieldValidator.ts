import { Path } from '../FieldRegister';
import { selectDeep, SelectDeepArgs } from '../selectDeep';
import { FieldStatus } from '../statusTracking/FieldStatus';
import { ArrayValidation, ValidationDefinition, ValidationFunction, ValidationResolver } from './ValidationDefinition';

export interface ValidateFieldArguments<Model> {
  model: Model;
  validationDefinition: ValidationDefinition<Model>;
  path: Path;
}

export class FieldValidator<Model> {
  private path: Path;
  private value: any;
  private model: Model;

  public static getValidationStatus<Model>(args: ValidateFieldArguments<Model>): Partial<FieldStatus> {
    const validator = new FieldValidator<Model>();
    return validator.getValidationStatus(args);
  }

  private getValidationStatus(args: ValidateFieldArguments<Model>): Partial<FieldStatus> {
    const error = this.validateField(args);
    const isValid = error === null;
    return {
      valid: isValid,
      inValid: !isValid,
      error: error || undefined
    };
  }

  private validateField({ model, validationDefinition, path }: ValidateFieldArguments<Model>): string | null {
    this.path = path;
    this.value = selectDeep({ object: model, path, assert: false });
    this.model = model;
    const validation: ValidationResolver = selectDeepValidator({ object: validationDefinition, path, assert: false });
    if (!validation) {
      return null;
    } else if (validation instanceof ArrayValidation) {
      return this.runArrayValidation(validation);
    } else {
      return this.runValidationFunctionInTryCatchAndCheckType(validation);
    }
  }

  private runArrayValidation(validation: ArrayValidation): string | null {
    if (this.value instanceof Array) {
      return this.runValidationFunctionInTryCatch(validation.arrayValidation);
    } else {
      return this.runValidationFunctionInTryCatchAndCheckType(validation.itemValidation);
    }
  }

  private runValidationFunctionInTryCatchAndCheckType(validationFunction: any): string | null {
    if (typeof validationFunction === 'function') {
      return this.runValidationFunctionInTryCatch(validationFunction);
    } else {
      const pathAsString = JSON.stringify(this.path);
      const errorMessage = `Invalid validation type "${typeof validationFunction}" for path "${pathAsString}"`;
      console.error(errorMessage);
      return errorMessage;
    }
  }

  private runValidationFunctionInTryCatch(validationFunction: ValidationFunction | undefined | null): string | null {
    try {
      return this.runValidationFunctionIfDefined(validationFunction);
    } catch (e) {
      const value = JSON.stringify(this.value);
      const path = JSON.stringify(this.path);
      const errorMessage = `Error while running validation function for value ${value} in path ${path}:`;
      console.error(errorMessage, e);
      return errorMessage;
    }
  }

  private runValidationFunctionIfDefined(validationFunction: ValidationFunction | undefined | null): string | null {
    if (validationFunction == null) {
      return null;
    }
    return validationFunction({ value: this.value, formValue: this.model }) || null;
  }
}

function selectDeepValidator({ object, path }: SelectDeepArgs): any {
  return path.reduce((item, key: string) => {
    if (item === undefined) {
      return;
    }
    if (item instanceof ArrayValidation) {
      return item.itemValidation;
    }
    return item[key];
  }, object);
}
