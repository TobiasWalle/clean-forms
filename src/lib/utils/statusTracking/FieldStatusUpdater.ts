import { FieldRegister, Path } from '../FieldRegister';
import { cloneFieldStatus, DEFAULT_FIELD_STATUS, FieldStatusArguments } from './FieldStatus';
import { FieldStatusMapping } from './FieldStatusMapping';

export class FieldStatusUpdater {
  constructor(
    private fieldRegister: FieldRegister
  ) {
  }

  public markAsDirty(status: FieldStatusMapping, path: Path): FieldStatusMapping {
    return this.updateField(status, path, {dirty: true});
  }

  public markAsTouched(status: FieldStatusMapping, path: Path): FieldStatusMapping {
    return this.updateField(status, path, {touched: true});
  }

  private updateField(
    status: FieldStatusMapping,
    path: Path,
    update: Partial<FieldStatusArguments>
  ): FieldStatusMapping {
    if (!this.fieldRegister.includesPath(path)) {
      return status;
    }
    const oldStatus = status[path];
    if (!oldStatus) {
      throw new ReferenceError(`Couldn't find path "${path}" in status mapping ${JSON.stringify(status)}`);
    }
    return { ...status, [path]: cloneFieldStatus(oldStatus, update) };
  }

  public addIfFieldNotExists(status: FieldStatusMapping, path: Path): FieldStatusMapping {
    if (status[path] == null) {
      return {
        ...status,
        [path]: DEFAULT_FIELD_STATUS
      };
    } else {
      return status;
    }
  }

  public removeIfFieldExists(status: FieldStatusMapping, path: Path): FieldStatusMapping {
    const copy = { ...status };
    delete copy[path];
    return copy;
  }
}
