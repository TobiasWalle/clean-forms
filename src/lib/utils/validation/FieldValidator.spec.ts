import { FieldStatus } from '../statusTracking/FieldStatus';
import { FieldValidator, ValidateFieldArguments } from './FieldValidator';
import { ArrayValidation } from './ValidationDefinition';

describe('FieldValidator', () => {
  describe('getValidationStatus', () => {
    it('should work if invalid', () => {
      const model = {
        a: 123,
        b: 'test',
      };
      type Model = typeof model;
      const error = 'Value has to be smaller than 100';
      const args: ValidateFieldArguments<Model> = {
        model,
        validationDefinition: {
          a: ({ value }) => value < 100 ? null : error
        },
        path: ['a']
      };
      const expectedResult: Partial<FieldStatus> = {
        inValid: true,
        valid: false,
        error
      };

      const result = FieldValidator.getValidationStatus(args);

      expect(result).toEqual(expectedResult);
    });

    it('should work if valid', () => {
      const model = {
        a: 50,
        b: 'test',
      };
      type Model = typeof model;
      const errorMessage = 'Value has to be smaller than 100';
      const args: ValidateFieldArguments<Model> = {
        model,
        validationDefinition: {
          a: ({ value }) => value < 100 ? null : errorMessage
        },
        path: ['a']
      };
      const expectedResult: Partial<FieldStatus> = {
        inValid: false,
        valid: true
      };

      const result = FieldValidator.getValidationStatus(args);

      expect(result).toEqual(expectedResult);
    });

    it('should catch errors in validation function', () => {
      const model = {
        a: 50,
        b: 'test',
      };
      type Model = typeof model;
      const args: ValidateFieldArguments<Model> = {
        model,
        validationDefinition: {
          a: () => {
            throw new Error('Test');
          }
        },
        path: ['a']
      };
      const expectedResult: Partial<FieldStatus> = {
        inValid: true,
        valid: false,
        error: expect.any(String)
      };
      console.error = jest.fn();

      const result = FieldValidator.getValidationStatus(args);

      expect(result).toEqual(expectedResult);
      expect(console.error).toHaveBeenCalledTimes(1);
      const errorMessage = (console.error as jest.Mock).mock.calls[0][0];
      expect(errorMessage).toMatchSnapshot();
    });

    it('should check if validation is invalid for an item', () => {
      const model = { a: [], };
      type Model = typeof model;
      const args: ValidateFieldArguments<Model> = {
        model,
        validationDefinition: { a: {} },
        path: ['a']
      };
      const expectedResult: Partial<FieldStatus> = {
        inValid: true,
        valid: false,
        error: expect.any(String)
      };
      console.error = jest.fn();

      const result = FieldValidator.getValidationStatus(args);

      expect(result).toEqual(expectedResult);
      expect(console.error).toHaveBeenCalledTimes(1);
      const errorMessage = (console.error as jest.Mock).mock.calls[0][0];
      expect(errorMessage).toMatchSnapshot();
    });

    it('should validate objects in an array', () => {
      interface Item {
        a: number;
        b: string;
      }

      interface Model {
        array: Item[];
      }

      const model: Model = { array: [{ a: 1, b: 'hello' }] };
      const error = 'Value has to be greater than 1';
      const validationDefinition = {
        array: new ArrayValidation<Model, Item>(
          { a: ({ value }) => value > 1 ? null : error },
        )
      };
      const args: ValidateFieldArguments<Model> = { model, validationDefinition, path: ['array', '0', 'a'] };
      const expectedResult: Partial<FieldStatus> = {
        inValid: true,
        valid: false,
        error,
      };

      const result = FieldValidator.getValidationStatus(args);

      expect(result).toEqual(expectedResult);
    });

    it('should validate numbers in an array', () => {
      interface Model {
        array: number[];
      }

      const model: Model = { array: [0, 1] };
      const error = 'Value has to be greater than 0';
      const validationDefinition = {
        array: new ArrayValidation<Model, number>(
          ({ value }) => value > 0 ? null : error
        )
      };
      const args: ValidateFieldArguments<Model> = { model, validationDefinition, path: ['array', '0'] };
      const expectedResult: Partial<FieldStatus> = {
        inValid: true,
        valid: false,
        error,
      };

      const result = FieldValidator.getValidationStatus(args);

      expect(result).toEqual(expectedResult);
    });

    it('should not validate objects in an empty array', () => {
      interface Item {
        a: number;
        b: string;
      }

      interface Model {
        array: Item[];
      }

      const model: Model = { array: [] };
      const error = 'Value has to be greater than 1';
      const validationDefinition = {
        array: new ArrayValidation<Model, Item>(
          { a: ({ value }) => value > 1 ? null : error },
        )
      };
      const args: ValidateFieldArguments<Model> = { model, validationDefinition, path: ['array'] };
      const expectedResult: Partial<FieldStatus> = {
        inValid: false,
        valid: true,
      };

      const result = FieldValidator.getValidationStatus(args);

      expect(result).toEqual(expectedResult);
    });

    it('should validate array itself', () => {
      interface Item {
        a: number;
        b: string;
      }

      interface Model {
        array: Item[];
      }

      const model: Model = { array: [{ a: 1, b: 'hello' }] };
      const error = 'Array has to be bigger than 1';
      const validationDefinition = {
        array: new ArrayValidation<Model, Item>(
          null,
          ({ value }) => value.length > 1 ? null : error
        )
      };
      const args: ValidateFieldArguments<Model> = { model, validationDefinition, path: ['array'] };
      const expectedResult: Partial<FieldStatus> = {
        inValid: true,
        valid: false,
        error,
      };

      const result = FieldValidator.getValidationStatus(args);

      expect(result).toEqual(expectedResult);
    });

    it('should run item validation if the value is not an array', () => {
      interface Model {
        value: number;
      }

      const model: Model = { value: 123 };
      const error = 'The number has to be smaller than 1';
      const validationDefinition = {
        value: new ArrayValidation<Model, number, any>(
          ({ value }) => value < 1 ? null : error
        )
      };
      const args: ValidateFieldArguments<Model> = { model, validationDefinition, path: ['value'] };
      const expectedResult: Partial<FieldStatus> = {
        inValid: true,
        valid: false,
        error,
      };

      const result = FieldValidator.getValidationStatus(args);

      expect(result).toEqual(expectedResult);
    });

    it('should validate nested arrays', () => {
      interface Item {
        a: number;
        b: string;
      }

      interface Model {
        array: Item[][][];
      }

      const model: Model = { array: [[[{ a: 1, b: 'hello' }]]] };
      const error = 'Value has to be bigger than 1';
      const validationDefinition = {
        array: new ArrayValidation(
          new ArrayValidation(
            new ArrayValidation<Model, Item>(
              { a: ({ value }) => value > 1 ? null : error },
            )),
        )
      };
      const args: ValidateFieldArguments<Model> = { model, validationDefinition, path: ['array', '0', '0', '0', 'a'] };
      const expectedResult: Partial<FieldStatus> = {
        inValid: true,
        valid: false,
        error,
      };

      const result = FieldValidator.getValidationStatus(args);

      expect(result).toEqual(expectedResult);
    });

    it('should work with nested paths', () => {
      const model = {
        a: { b: { c: 123 } },
      };
      type Model = typeof model;
      const error = 'Value has to be smaller than 100';
      const args: ValidateFieldArguments<Model> = {
        model,
        validationDefinition: {
          a: { b: { c: ({ value }) => value < 100 ? null : error } }
        },
        path: ['a', 'b', 'c']
      };
      const expectedResult: Partial<FieldStatus> = {
        inValid: true,
        valid: false,
        error
      };

      const result = FieldValidator.getValidationStatus(args);

      expect(result).toEqual(expectedResult);
    });
  });
});
