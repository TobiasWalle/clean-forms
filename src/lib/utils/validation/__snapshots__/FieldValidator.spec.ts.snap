// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FieldValidator getValidationStatus should catch errors in validation function 1`] = `"Error while running validation function for value 50 in path [\\"a\\"]:"`;

exports[`FieldValidator getValidationStatus should check if validation is invalid for an item 1`] = `"Invalid validation type \\"object\\" for path \\"[\\"a\\"]\\""`;
